theano/scan_module/scan_opt.py

```diff
@@ -294,6 +326,14 @@ def add_to_replace(y):
                 # We have a candidate node to removable
                 # Step 1. Reconstruct it on outside
                 to_remove_set.add(nd)
+                node_clients = retrieve_clients_fromnodes(nd)
+
+                # Add the nodes only if they are not already
+                # in the bookkeeper's list
+                nd_cli_cleaned = filter(lambda x: x not in local_fgraph_bookkeeper,
+                                        node_clients)
+                local_fgraph_bookkeeper.extend(nd_cli_cleaned)
```

Comment:
Also, now you are adding to the right of your deque and then popping, also from the right. This amounts to doing depth-first search. You should add to the left and pop from the right => this will amount to breadth-first search